src/include/OpenImageIO/thread.h

```diff
@@ -601,14 +616,33 @@ OIIO_API thread_pool* default_thread_pool ();
 template<typename T>
 class task_set {
 public:
-    task_set () { }
+    task_set (thread_pool *pool) { m_pool = pool; }
     ~task_set () { wait(); }
     void push (std::future<T> &&f) { m_futures.emplace_back (std::move(f)); }
     void wait () {
-        for (auto& f : m_futures)
-            f.wait();
+        int tries = 0;
+        std::chrono::milliseconds wait_time (0);
+        while (1) {
+            bool all_finished = true;
+            for (auto& f : m_futures) {
+                // Asking future.wait_for for 0 time just checks the status.
+                auto status = f.wait_for (wait_time);
+                if (status != std::future_status::ready)
+                    all_finished = false;
+            }
+            if (all_finished)   // All futures are ready? We're done.
+                break;
+            // We're still waiting on some tasks to complete. What next?
+            if (++tries < 4)    // First few times,
+                continue;       //   just busy-wait, check status again
+            // Since we're waiting, try to run a task ourselves to help
```

Comment:
I updated the PR with this change. The default (or if you explicitly pass nthreads=0 to the constructor or to resize) now means "number of cores - 1".  If you pass a nonzero number of cores, it takes your request literally and it's up to you to size the pool (or adjust your use of it) to avoid over-threading. I also added an optional parameter to task_set::wait() that lets you choose to do a true wait and NOT pick up any of the work yourself, but the default is to contribute to the work rather than blocking.